### ISSUE TITLE
Clock is not immediately displayed in the user's local time

### ISSUE THREAD

**Issue by tutaleedoo** [QubesOS/qubes-issues#3374]
#### Qubes OS version:

Qubes release 4.0 (R4.0)

#### Affected TemplateVMs: 

dom0

---

### Steps to reproduce the behaviour:

1. Start Qubes OS

### Expected behaviour:

- The clock displays in the user's local time

### Actual behaviour:

- The clock displays in UTC

### General notes:

- Eventually, I don't know how, the clock displays in the user's local time
- It doesn't matter whether the clock's timezone settings are set or not (right-click the clock, click `Properties`, see `Timezone:`)
- After some seemingly random use of Qubes OS, something automatically causes the clock to appear in the user's local time
- dom0 autoconnects to the network on startup, so does have a network connection

---

**Comment by andrewdavidwong** [comment]
This is possibly related to (but does not appear to be a duplicate of) #3210.

As for how the time system works, Marek explained it [here](https://groups.google.com/d/msg/qubes-devel/2KRLCfdhlwE/vljpHMWwBgAJ):
> Here is how it works: 
> 
> 1. sys-net runs some standard clock synchronization service (ntpd, 
> systemd-timesyncd - depending on template); it is enabled by setting 
> 'clocksync' service (qvm-service tool in dom0). 
> 
> 2. Then every VM, including dom0, synchronize its time directly from 
> there. In case of VMs, it is done by qvm-sync-clock (called from 
> qubes-time-sync.service in the VM), controlled by two things: 
>   - absence (or disabling) of 'clocksync' service (qvm-service) 
>   - qrexec policy, where actual VM used for time sync is set 
>     (/etc/qubes-rpc/policy/qubes.GetDate) 
> For dom0, also qvm-sync-clock is used (slightly different one for dom0), 
> and the VM is set by 'clockvm' global setting (qubes-prefs). 
> 
> In theory, all those places should synchronize UTC time, and honour 
> local timezone for display purposes. But maybe somewhere it is messed 
> up... I'd start with checking timezone in sys-net and time 
> synchronization with the network (systemd-timesyncd.service) there.

**Comment by nunesgh** [comment]
Same issue here. But in my case, I've also noticed that, when the clock automatically updates to my local time, it doesn't happen to all VMs. Right now, sys-net, dom0 (apparently), and untrusted appear to be in the correct time, but personal is still in UTC.

**Comment by nunesgh** [comment]
Marek has suggested, on the mailing list you linked above, to try calling `sudo hwclock --systohc --utc`, but when I do so, I receive the message:

> hwclock: Cannot access the Hardware Clock via any known method.
> hwclock: Use the --debug option to see the details of our search for an access method.

When using the debug option, it returns:

> hwclock from util-linux 2.28.2
> hwclock: cannot open /dev/rtc: No such file or directory
> No usable clock interface found.
> hwclock: Cannot access the Hardware Clock via any known method.

I hope this might help to find the source of the issue!

**Comment by tick10** [comment]
I don't know what the source of this issue is but i was experiencing the same thing after a fresh install of qubes 4. I was also having the added issue of my sys-whonix vm being unable to connect to tor because the time/date was more than 30 minutes off. 

I did however find a workaround that seems to be working fine for me so far. If you set your bios/hardware clock manually (in your bios/UEFI) to utc time instead of your local time (whatever that happens to be) then the clock seems to be set correctly on boot after that. This has worked for me at least.

**Comment by marmarek** [comment]
It should be set to UTC. So in fact, it looks like #4266

**Comment by andrewdavidwong** [comment]
> It should be set to UTC. So in fact, it looks like #4266

@marmarek, are you saying this is a duplicate of #4266?

**Comment by gw0** [comment]
> Right now, sys-net, dom0 (apparently), and untrusted appear to be in the correct time, but personal is still in UTC.

I am experiencing same symptoms in R4.1-beta1. Debugging:

- my local time zone is CEST (UTC+2), but I have BIOS/RTC is set to UTC (as recommended by `timedatectl`)
- disconnect from ethernet/wifi to prevent NTP synchronization and reboot
- *dom0* after a fresh installation it is using RTC in local TZ (dom0 and AppVM times are off until you NTP synchronizes them), to set RTC to UTC use: `timedatectl set-local-rtc 0` (no more issues with dom0)
- *dom0* is **correct** right after boot with no net:
    ```
$ timedatectl status
               Local time: Tue 2021-08-10 08:30:31 CEST
           Universal time: Tue 2021-08-10 06:30:31 UTC
                 RTC time: n/a
                Time zone: Europe/Ljubljana (CEST, +0200)
System clock synchronized: no
              NTP service: inactive
          RTC in local TZ: no
    ```
- *dom0* remains correct even after connected to the net for a while
- *personal* AppVM (debian-10) is **incorrect** right after boot with no net:
    ```
               Local time: Tue 2021-08-10 10:05:21 CEST
           Universal time: Tue 2021-08-10 08:05:21 UTC
                 RTC time: n/a
                Time zone: Europe/Ljubljana (CEST, +0200)
System clock synchronized: no
              NTP service: inactive
          RTC in local TZ: no
    ```
- in the same *personal* AppVM remains incorrect even after connected to the net for a while
- after restarting *personal* AppVM it is **correct** (connected to net):
    ```
               Local time: Tue 2021-08-10 08:22:21 CEST
           Universal time: Tue 2021-08-10 06:22:21 UTC
                 RTC time: n/a
                Time zone: Europe/Ljubljana (CEST, +0200)
System clock synchronized: no
              NTP service: inactive
          RTC in local TZ: no
    ```
- *untrusted* AppVM (debian-10) is **correct** right after boot with no net:
    ```
               Local time: Tue 2021-08-10 08:31:52 CEST
           Universal time: Tue 2021-08-10 06:31:52 UTC
                 RTC time: n/a
                Time zone: Europe/Ljubljana (CEST, +0200)
System clock synchronized: no
              NTP service: inactive
          RTC in local TZ: no
    ```
- the same *untrusted* AppVM becomes **incorrect** after connected to the net for a while:
    ```
               Local time: Tue 2021-08-10 10:45:11 CEST
           Universal time: Tue 2021-08-10 08:45:11 UTC
                 RTC time: n/a
                Time zone: Europe/Ljubljana (CEST, +0200)
System clock synchronized: no
              NTP service: inactive
          RTC in local TZ: no
    ```
- after restarting *untrusted* AppVM it is **correct** (connected to net):
   ```
               Local time: Tue 2021-08-10 08:49:11 CEST
           Universal time: Tue 2021-08-10 06:49:11 UTC
                 RTC time: n/a
                Time zone: Europe/Ljubljana (CEST, +0200)
System clock synchronized: no
              NTP service: inactive
          RTC in local TZ: no
   ```
- a brand-new AppVM based on debian-10 behaves like *untrusted*

I believe this issue is not experienced so often, because most users are always connected to the net and NTP synchronization corrects the time in most cases.

**Comment by DemiMarie** [comment]
What are your virtualization modes?  Handling of timezones in HVM mode is known to be buggy; see https://github.com/QubesOS/qubes-core-admin/pull/417 for a fix.  As a workaround, you can set the `timezone` feature of all templates to the empty string.

**Comment by gw0** [comment]
> What are your virtualization modes?

Default (PVH).

> As a workaround, you can set the `timezone` feature of all templates to the empty string.

Thank you, everything works correctly now (right after boot and even after connected to the net). To sum up the steps needed to set up RTC in UTC:

```
## in dom0:
$ timedatectl set-local-rtc 0
$ qvm-features debian-10 timezone ''
## reboot
```

**Comment by qtpies** [comment]
> ```
> $ qvm-features debian-10 timezone ''
> ## reboot
> ```

Setting this led to some, not all, of my HVM qubes not starting anymore. I didnt see an obvious difference. Nothing is actually added to the libxl-driver.log file. I think this is related to [this issue](https://github.com/QubesOS/qubes-issues/issues/5616) which mentions it will be fixed in 4.1. So watch out applying this workaround until 4.1 is out. 

```
[user@dom0 ~]$ qvm-start -v sys-net
Start failed: unsupported configuration: unknown clock basis 'UTC', see /var/log/libvirt/libxl/libxl-driver.log for details
```

**Comment by DemiMarie** [comment]
> > ```
> > $ qvm-features debian-10 timezone ''
> > ## reboot
> > ```
> 
> Setting this led to some, not all, of my HVM qubes not starting anymore. I didnt see an obvious difference. Nothing is actually added to the libxl-driver.log file. I think this is related to [this issue](https://github.com/QubesOS/qubes-issues/issues/5616) which mentions it will be fixed in 4.1. So watch out applying this workaround until 4.1 is out.

This was a workaround for a (now fixed) R4.1 bug, and was never necessary in R4.0 if I recall correctly.

**Comment by qtpies** [comment]
> > > ```
> > > $ qvm-features debian-10 timezone ''
> > > ## reboot
> > > ```
> > 
> > 
> > Setting this led to some, not all, of my HVM qubes not starting anymore. I didnt see an obvious difference. Nothing is actually added to the libxl-driver.log file. I think this is related to [this issue](https://github.com/QubesOS/qubes-issues/issues/5616) which mentions it will be fixed in 4.1. So watch out applying this workaround until 4.1 is out.
> 
> This was a workaround for a (now fixed) R4.1 bug, and was never necessary in R4.0 if I recall correctly.

Ok, I see, I didn't see that. I tried this as I'm desperate to fix my timesync issues, it is the thing bothering me most about Qubes, and it took me an hour to fix it, so I mainly wanted to document this somewhere for others encountering this error.

**Comment by larryswinger** [comment]
Is there a fix for Qubes 4.0 users yet that doesn't break the system?

**Comment by andrewdavidwong** [comment]
Sounds like this was fixed in 4.1. Closing as resolved. If anyone believes this issue is not yet resolved, or if anyone is still affected by this issue, please leave a comment, and we'll be happy to reopen it. Thank you.
